bftengine/src/bftengine/ReplicaForStateTransfer.cpp

```diff
@@ -70,18 +73,44 @@ ReplicaForStateTransfer::ReplicaForStateTransfer(const ReplicaConfig &config,
 }
 
 void ReplicaForStateTransfer::start() {
-  cre_ = bftEngine::bcst::asyncCRE::CreFactory::create(msgsCommunicator_, msgHandlers_);
+  /* TODO: Transaction signing needs to be disabled for replicas, as state transfer may change
+   * the a replica's consensus (and therefore main) key. Apollo tests currently run with transaction signing enabled,
+   * we therefore use the latest key published by this replica.
+   * If a key exchange consensus was reached by this replica, the other replicas are guaranteed to be able to use
+   * its latest key upon honest execution of the exchange */
+  cre_ = bftEngine::bcst::asyncCRE::CreFactory::create(
+      msgsCommunicator_, msgHandlers_, std::make_unique<bftEngine::bcst::asyncCRE::ReplicaCRESigner>());
   stateTransfer->setReconfigurationEngine(cre_);
   stateTransfer->addOnTransferringCompleteCallback(
-      [this](std::uint64_t) {
+      [this](std::uint64_t checkpoint) {
         // TODO - The next lines up to comment 'YYY' do not belong here (CRE) - consider refactor or move outside
         if (!config_.isReadOnly) {
+          // Load the public keys of the other replicas from reserved pages
+          // so that their responses can be validated
+          KeyExchangeManager::instance().loadPublicKeys();
+
+          // Make sure to sign the reconfiguration client messages using the key
+          // other replicas expect
+          SigManager::instance()->setReplicaLastExecutedSeq(checkpoint * checkpointWindowSize);
+
+          // Need to update private key to match the loaded public key in case they differ (key exchange was executed
+          // on other replicas but not on this one, finishing ST does not mean that missed key exchanges are executed)
+          // This can be done by iterating the saved cryptosystems and updating their private key if their
+          // public key matches the candidate saved in KeyExchangeManager
+          // TODO: persist the candidate
+          CryptoManager::instance().onCheckpoint(checkpoint);
+          auto [priv, pub] = KeyExchangeManager::instance().getCandidateKeyPair();
```

Comment:
Notice that this PR https://github.com/vmware/concord-bft/pull/2932 enables multiple candidates. 
So (1) the getCandidateKeyPair should be changed accordingly (maybe return all the existing candidates) and (2) you need to try and match the correct candidate with the one that we have on the reserved pages.

Another general comment:
https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/src/bftengine/KeyExchangeManager.cpp#L98
does multiple things (like clearing the candidate), persisting the private key, and more. Maybe, it is worth simply calling it directly here such that we won't miss anything.

Reply:
Since the candidate is not persisted, I thought that the persistency should have been handled in a separate PR.
Nevertheless, I've implemented the logic you requested.